pkg/internalapis/cnsvolumeoperationrequest/cnsvolumeoperationrequest.go

```diff
@@ -122,20 +108,130 @@ func InitVolumeOperationRequestInterface(ctx context.Context) (VolumeOperationRe
+	// last operation that was invoked on a volume.
+	var operationDetailsToReturn cnsvolumeoperationrequestv1alpha1.OperationDetails
+	if len(instance.Status.LatestOperationDetails) != 0 {
+		operationDetailsToReturn = instance.Status.LatestOperationDetails[len(instance.Status.LatestOperationDetails)-1]
+	} else {
+		operationDetailsToReturn = instance.Status.FirstOperationDetails
+	}
+
+	return CreateVolumeOperationRequestDetails(instance.Spec.Name, instance.Status.VolumeID, instance.Status.SnapshotID,
+			instance.Status.Capacity, operationDetailsToReturn.TaskInvocationTimestamp, operationDetailsToReturn.TaskID,
+			operationDetailsToReturn.OpID, operationDetailsToReturn.TaskStatus, operationDetailsToReturn.Error),
+		nil
 }
 
 // StoreRequestDetails persists the details of the operation taking
 // place on the volume by storing it on the API server.
 // Returns an error if any error is encountered. Clients must assume
 // that the attempt to persist the information failed if an error is returned.
-func (or *operationRequestStoreOnETCD) StoreRequestDetails(ctx context.Context, instance *VolumeOperationRequestDetails) error {
+func (or *operationRequestStoreOnETCD) StoreRequestDetails(ctx context.Context, operationToStore *VolumeOperationRequestDetails) error {
 	log := logger.GetLogger(ctx)
-	// TODO: Implement StoreRequestDetails for operationRequestStoreOnETCD
-	err := fmt.Errorf("StoreRequestDetails not implemented for operationRequestStoreOnETCD")
-	log.Error(err)
-	return err
+	log.Debugf("Storing CnsVolumeOperationRequest instance with spec %v", spew.Sdump(operationToStore))
+
+	operationDetailsToStore := convertToCnsVolumeOperationRequestDetails(*operationToStore.OperationDetails)
+
+	instance := &cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequest{}
+	log.Infof("Getting CnsVolumeOperationRequest instance with name %s/%s", csiconfig.DefaultCSINamespace, operationToStore.Name)
+
+	namespacedName := client.ObjectKey{Name: operationToStore.Name, Namespace: csiconfig.DefaultCSINamespace}
+	if err := or.k8sclient.Get(ctx, namespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			// Create new instance on API server if it doesnt exist. Implies that this is the first time this object is being stored.
+			newInstance := &cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      namespacedName.Name,
+					Namespace: namespacedName.Namespace,
+				},
+				Spec: cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequestSpec{
+					Name: namespacedName.Name,
+				},
+				Status: cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequestStatus{
+					VolumeID:               operationToStore.VolumeID,
+					SnapshotID:             operationToStore.SnapshotID,
+					Capacity:               operationToStore.Capacity,
+					FirstOperationDetails:  *operationDetailsToStore,
+					LatestOperationDetails: nil,
```

Comment:
Why dont you set the LatestOperationDetails too? After all it is supposed to hold the last 'x' number of operations.

Reply:
Done